### PR TITLE
feat: [SMP-2324]: Add template function for podmonitors in helm-common

### PR DESCRIPTION
**Pull request by sarthak-harnessio**
GMP scrapes metrics through K8s CRDs (Custom Resource Definitions). We are using PodMonitoring CRD to scrape metrics from all our clusters. The existing the podMonitoring yaml below -

QA - https://github.com/wings-software/ng-prod-manifests/blob/quality-assurance/pod-monitoring/pl-podMonitors.yaml 

Prod - 

https://github.com/wings-software/ng-prod-manifests/blob/master/environments/prod/pod-monitoring/pl-podMonitors.yaml 

https://github.com/wings-software/ng-prod-manifests/blob/master/environments/free/pod-monitoring/pl-podMonitors.yaml 

https://github.com/wings-software/ng-prod-manifests/blob/master/environments/compliance/pod-monitoring/pl-podMonitors.yaml 

So these PodMonitor CRDs have to be maintained for qa0, prod0 and harness0 and have to be deployed in an automated fashion whenever updates are made.

### CODE REVIEW

File: src/common/templates/_monitoring.tpl [hunk 1]
```diff
@@ -43,3 +43,31 @@ prometheus.io/scrape: {{ $enabled | quote}}
...
+{{- if $enabled -}}
+apiVersion: monitoring.googleapis.com/v1
+kind: PodMonitoring
+metadata:
+  name: {{ .name }}
+  namespace:  {{ $namespace }}
+spec:
+  selector:
+    matchLabels:
+      app: {{ .name }}
```
Comment (kapilgarg1996) [inline src/common/templates/_monitoring.tpl:67]: lets use the labels which are being used in deployment of each service. we can take these labels as arguments as well.

File: src/common/templates/_monitoring.tpl [hunk 2]
```diff
@@ -43,3 +43,31 @@ prometheus.io/scrape: {{ $enabled | quote}}
...
+{{- $globalMonitoring := default (dict) ((pluck "monitoring" .Values.global) | first) -}}
+{{- $monitoring := (mergeOverwrite $globalMonitoring $localMonitoring ) }}
+{{- $port := (pluck "port" $monitoring) | first }}
+{{- $path := (pluck "path" $monitoring) | first }}
+{{- $namespace := $.Release.Namespace }}
+{{- if $enabled -}}
+apiVersion: monitoring.googleapis.com/v1
+kind: PodMonitoring
+metadata:
+  name: {{ .name }}
```
Comment (kapilgarg1996) [inline src/common/templates/_monitoring.tpl:62]: this can also be default to `$.Chart.Name`

File: src/common/templates/_monitoring.tpl [hunk 3]
```diff
@@ -43,3 +43,31 @@ prometheus.io/scrape: {{ $enabled | quote}}
   protocol: "TCP"
 {{- end }}
 {{- end -}}
+
+{{/* Podmonitor template to be added for Google Managed prometheus
```
Comment (akshay-khole) [inline src/common/templates/_monitoring.tpl:47]: so all services/ charts will use this template function for monitoring ? how are we making sure smp customers will not be affected by this change ?

**Comment by arya-harness** [comment]
update chart.yaml and increment version

File: src/common/templates/_monitoring.tpl [hunk 4]
```diff
@@ -43,3 +43,32 @@ prometheus.io/scrape: {{ $enabled | quote}}
...
+{{- if $enabled -}}
+apiVersion: monitoring.googleapis.com/v1
+kind: PodMonitoring
+metadata:
+  name: {{ $.Chart.Name }}
+  namespace:  {{ $namespace }}
+spec:
+  selector:
+    matchLabels:
+      app: {{ default $.Chart.Name .name }}
```
Comment (kapilgarg1996) [inline src/common/templates/_monitoring.tpl:68]: this should be same as the one being used by the service's Deployment. Also the whole code should either use `.ctx` or `$` as both mean the same thing above.

File: src/common/Chart.yaml [hunk 5]
```diff
@@ -15,12 +15,15 @@ type: library
...
 # follow Semantic Versioning. They should reflect the version the application is using.
 # It is recommended to use it with quotes.
 appVersion: "0.0.0"
-
-dependencies:
+description: Helm Chart for deploying Harness.
+name: harness
+type: application
+version: 0.11.0
+dependencies:
```
Comment (avaakash) [inline src/common/Chart.yaml:29]:
```suggestion
```

File: src/common/Chart.yaml [hunk 6]
```diff
@@ -15,12 +15,15 @@ type: library
...
+version: 1.3.1
 
 # This is the version number of the application being deployed. This version number should be
 # incremented each time you make changes to the application. Versions are not expected to
 # follow Semantic Versioning. They should reflect the version the application is using.
 # It is recommended to use it with quotes.
 appVersion: "0.0.0"
-
-dependencies:
+description: Helm Chart for deploying Harness.
```
Comment (avaakash) [inline src/common/Chart.yaml:25]: Why is this being changed?

**Pull request closed without merging**